### ISSUE TITLE
If a nupkg/nuspec is corrupt, do not fail trying to read it

### ISSUE THREAD

**Issue by ferventcoder** [chocolatey/choco#717]
There may have been a network issue.  Related to #198.

Zero byte nupkgs or nuspec files will give an issue like "Root element is missing". 

### Workaround

Find and remove the zero byte file and that should fix the issue. This is a manual workaround until Chocolatey can handle this.

**Comment by ferventcoder** [comment]
@DarwinJS can you get me details on this one again? I just attempted to recreate this and was unable to.

**Comment by DarwinJS** [comment]
If I remember correctly the file was truncated and the install failed.
Then when trying to install again it would not redownload (because the file
passes an existence check?), but then would choke on the incomplete file.
So kind of a catch 22 if you don't know what to delete.  Maybe it is
important that the type of file corruption is a truncation?

D.

**Comment by majkinetor** [comment]
I got this with completely empty nuspec file.

**Comment by zertyz** [comment]
The same just happened for me, as described by @majkinetor. Why postpone this integrity checking, when having such an error will make choco unusable for the average user?

**Comment by ferventcoder** [comment]
> Why postpone this integrity checking, when having such an error will make choco unusable for the average user?

Careful, it's not a postpone. It's more of it never existed kind of thing. This is about adding it, but also that we are trying to determine where the work is for this.

**Comment by marcinbojko** [comment]
Hit by the same issue - out of space on one of dev servers, downloaded nupkg/nuspec files were broken (0 in lenght).

**Comment by masgo** [comment]
I suggest starting with a more useful error message. "Root element is missing" is not very helpful. Maybe make it something like: "Root element is missing. Probably the nupkg file is broken. You can try to delete it by (add some command to chocolatey which just deletes everything)  and start over." This would solve the issues for many users. 

Later™, you could do a more sophisticaed/automated solution.

**Comment by ferventcoder** [comment]
@masgo I completely agree with better/more actionable errors - it's a constant struggle when you use libraries that don't share this same idea.

**Comment by ferventcoder** [comment]
Note for the implementer - take a look at https://github.com/chocolatey/choco/issues/1829 as well.

**Comment by jawn** [comment]
To make it easier to discover, I documented the issue and workaround [on superuser.com.](https://superuser.com/questions/1553128/error-message-root-element-is-missing/1553129)

**Comment by ferventcoder** [comment]
Thanks for doing that @jawn!

**Comment by scls19fr** [comment]
For those who also encounter this zero size file bug, you can remove them using this PowerShell command.

    ChildItem C:\ProgramData\chocolatey\lib\*\*.nu* | where Length -eq 0 | Del

**Comment by Chirishman** [comment]
Ran into this again today when trying to upgrade the mediainfo package, kept getting

`2022-02-08 04:29:20,338 18004 [ERROR] - chocolatey not upgraded. An error occurred during installation:
 Unable to read package from path 'imagemagick.app\imagemagick.app.nupkg'.`

Then when I removed the imagemagick.app folder from lib I started getting "root element is missing" until I ran @scls19fr's cleanup script.

+1 to this being handled more gracefully

**Comment by pauby** [comment]
I'm going to close this as a duplicate of #3124 only because it uses our template and is more up-to-date.
